src/app/digital/utils/ExpressionParser/GenerateInputTree.ts

```diff
@@ -0,0 +1,144 @@
+    index += 1;
+    if (index >= tokens.length && currentOp !== "(") {
+        throw new Error("Missing Right Operand: \"" + ops.get(currentOp) + "\"");
+    }
+    let rightRet: NewTreeRetValue = null;
+    const rightToken = tokens[index];
+    if (currentOp === "!" && rightToken.type === "!") // This case applies when there are two !'s in a row
+        rightRet = generateInputTreeCore(tokens, ops, currentOp, index);
+    else if (currentOp === "!" && rightToken.type === "input") // This case would apply when an input follows a "!"
+        rightRet = {index: index+1, tree: {kind: "leaf", ident: rightToken.name}};
+    else if (currentOp === "(") {
+        if (index >= tokens.length)
+            throw new Error("Encountered Unmatched \"" + ops.get("(") + "\"");
+        rightRet = generateInputTreeCore(tokens, ops, nextOp, index);
+    }
+    else
+        rightRet = generateInputTreeCore(tokens, ops, currentOp, index);
+    index = rightRet.index;
+    if (currentOp === "(") {
+        if (index >= tokens.length)
+            throw new Error("Encountered Unmatched \"" + ops.get("(") + "\"");
+        if (tokens[index].type !== ")")
+            throw new Error("Encountered Unmatched \"" + ops.get("(") + "\"");
+        rightRet.index += 1; // Incremented to skip the ")"
+        return rightRet;
+    }
+
+    // The tree tree is created with the new node as the root and returned
+    let tree: InputTree;
+    if (currentOp === "!")
+        tree = {kind: "unop", type: "!", child: rightRet.tree};
+    else if (currentOp === "|" || currentOp === "^" || currentOp === "&")
+        tree = {kind: "binop", type: currentOp, lChild: leftRet.tree, rChild: rightRet.tree};
+    return {index: index, tree: tree};
+
+}
+
+/**
+ * The core of the function to generate the input tree. Various errors are returned for invalid inputs
+ * 
+ * @param tokens the array of tokens representing the expression to parse
+ * @param ops the representation format for the operations used in this expression (only used for error messages)
+ * @returns null if tokens.length is 0, the relevant input tree otherwise
+ * @throws {Error} parenthesis do not include anything (such as "()")
+ * @throws {Error} an opening parenthesis is missing a corresponding closing parenthesis (such as "(")
+ * @throws {Error} a closing parenthesis is missing a corresponding opening parenthesis (such as ")")
+ * @throws {Error} |, &, or ^ are missing an operand on their left (such as "a|")
+ * @throws {Error} |, &, ^, or ! are missing an operand on their right (such as "!a")
+ * @throws {Error} there is no operator between two inputs (such as "a b")
+ * @throws {Error} generateInputTreeCore returns back up to this function before the end of tokens is reached
+ *                  for any other reason
+ */
+export function GenerateInputTree(tokens: Array<Token>, ops: Map<FormatLabels, string> = FormatMap.get("|")): InputTree | null {
```

Comment:
In general see if any of these for loops can be reduced to like `map`, `find`, `some`, `every`